runtime/src/accounts_index.rs

```diff
@@ -611,22 +943,40 @@ impl<T: 'static + Clone> AccountsIndex<T> {
         }
     }
 
+    fn purge_secondary_indexes_by_inner_key(
+        &self,
+        inner_key: &Pubkey,
+        slots_to_remove: Option<&HashSet<Slot>>,
+    ) {
+        self.program_id_index
```

Comment:
maybe bail out early if we aren't indexing like indexing path to avoid this meaningless loop over N where number of slots (3x empty lookup on every account write)?: https://github.com/solana-labs/solana/pull/14212/files#r548632636

Reply:
@ryoqun do you mean switch the indexes to Option types (instantiate based on validator args) and don't call these functions if those indexes are `None`?

Reply:
@carllin I meant simpler ones like this (similar to `update_secondary_indexes`):

```rust
if account_indexes.contains(&AccountIndex::ProgramId) {
    self.program_id_index.remove_by_inner_key(inner_key, slots_to_remove);
}
if account_indexes.contains(&AccountIndex::SplTokenMint) {
    self.spl_token_mint.remove_by_inner_key(inner_key, slots_to_remove);
}
```